### PR TITLE
Add the proposals for `protocol-cosmwasm`

### PR DESCRIPTION
**Pull request by duguorong009**
Implement the [cosmwasm-proposal spec](https://github.com/webb-tools/protocol-cosmwasm/issues/63)

### CODE REVIEW

File: proposals/src/proposal/cosmwasm/rescue_tokens.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+/// Rescue Tokens Proposal.
+///
+/// The `RescueTokensProposal` rescues tokens from the treasury to a specified
+/// `to` address.
+/// The format of the proposal looks like this:
+///
+/// ```text
+/// ┌────────────────────┬───────────────────┬────────────────────┬────────────┐
+/// │                    │                   │                    │            │
+/// │ ProposalHeader 40B │ TokenAddress 20B  │ ToTokenAddress 20B │ Amount 32B │
```
Comment (drewstone) [inline proposals/src/proposal/cosmwasm/rescue_tokens.rs:13]: `ToTokenAddress` should be `Recipient`

**Review by drewstone — CHANGES_REQUESTED** [review]
If these are cosmwasm proposals, can we also update the tests so that we compute TargetSystems and contract addresses the same way we plan to in CW? For example, we know contract addresses are larger than 20 and 32 bytes so we have to hash them. Can you provide that functionality to the tests? Right now, it looks mostly copied from EVM proposals.

Also, if these proposals are copied from EVM, please do not copy all the code. Simply re-export the proposal from this directory. If not, can you provide and describe what the changes are between proposals?

**Comment by duguorong009** [comment]
@drewstone 
I think we should discuss the following issues: 
- Some proposals are pretty big when converted to bytes
   For example, [wrapping_fee_update](https://github.com/webb-tools/webb-rs/blob/guorong/add-cosmwasm-proposals/proposals/src/proposal/cosmwasm/wrapping_fee_update.rs) is pretty big.
   The reason is that `cosmwasm` implementation uses the [large struct](https://github.com/webb-tools/webb-rs/blob/guorong/add-cosmwasm-proposals/proposals/src/proposal/cosmwasm/wrapping_fee_update.rs#L100-L106).
   Maybe, should I update the `protocol-cosmwasm`? so that it can just use the simple struct(one field)?
   
- All proposals include `proposal_header`. But, `protocol-cosmwasm` is not implemented for handling them.
   The `proposal_header` is `resource_id` + `func_sig` + `nonce`.
   But, current `cosmwasm` contracts regard the proposal as `resource_id` + `message`, not the `proposal_header` + `message`.
   Maybe, should I also update the `protocol_cosmwasm` for compatibility?
   I mean, to update the `cosmwasm` contracts so that they can handle the `proposal_header(40 bytes)` + `message`

File: proposals/src/proposal/cosmwasm/rescue_tokens.rs (hunk 1 above)
Comment (drewstone) [reply proposals/src/proposal/cosmwasm/rescue_tokens.rs:13]: Can you also change it in the EVM side please.

**Comment by drewstone** [comment]
> All proposals include proposal_header. But, protocol-cosmwasm is not implemented for handling them.
The proposal_header is resource_id + func_sig + nonce.
But, current cosmwasm contracts regard the proposal as resource_id + message, not the proposal_header + message.
Maybe, should I also update the protocol_cosmwasm for compatibility?
I mean, to update the cosmwasm contracts so that they can handle the proposal_header(40 bytes) + message

Please don't deviate from how the protocol is implemented. You should be able to answer this question, simply implement it as was stated in the task, which is to match the other protocols.

**Comment by drewstone** [comment]
Every question that relates "should I update protocol-cosmwasm to match protocol-solidity" and "substrate" when applicable should be immediately answered with YES unless there are clear differences such as the encoding and lengths of addresses.

**Comment by duguorong009** [comment]
> Every question that relates "should I update protocol-cosmwasm to match protocol-solidity" and "substrate" when applicable should be immediately answered with YES unless there are clear differences such as the encoding and lengths of addresses.
> 
> Since we also addressed that we would hash longer values and use them to fit the lengths of the current proposal fields, I would also expect to have that represented in tests of CosmWasm contracts. We should use actual cosmwasm contract addresses as well. Until then, this PR is not ready.

Updated the `protocol-cosmwasm` repo as requested. (https://github.com/webb-tools/protocol-cosmwasm/pull/66)
Also, updated the unit tests so that real addresses(`juno` network) are used for them.

File: proposals/src/proposal/cosmwasm/rescue_tokens.rs (hunk 1 above)
Comment (duguorong009) [reply proposals/src/proposal/cosmwasm/rescue_tokens.rs]: Yes, I can handle that change.
This is the PR I created for above recommend.
https://github.com/webb-tools/webb-rs/pull/65

File: proposals/src/proposal/cosmwasm/anchor_update.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    };
+
+    use super::*;
+
+    const TARGET_CONTRACT_ADDR: &str =
+        "juno1hset4pny4h8xm4s4lek57msq7j4zwfqwjf7zxqjt4npxyv0lrgnsp8qy9j";
+
+    #[test]
+    fn encode() {
+        let target_addr = cosmos_addr_2_target_addr(TARGET_CONTRACT_ADDR);
```
Comment (drewstone) [inline proposals/src/proposal/cosmwasm/anchor_update.rs:179]: Can we rename this to `cosmos_address_to_target_system`?

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks good, some changes I think would be good on the naming front.

File: proposals/src/proposal/cosmwasm/anchor_update.rs (hunk 2 above)
Comment (duguorong009) [reply proposals/src/proposal/cosmwasm/anchor_update.rs:179]: Um, atm, it is for converting the `cosmos address` to **20 bytes** hashed address, not the `target_system`(26 bytes).
That's why I named it as `cosmos_addr_2_target_addr`.

Comment (drewstone) [reply proposals/src/proposal/cosmwasm/anchor_update.rs:179]: Can you rename it to `cosmos_address_to_target_address` then please? I don't particularly like mixing numbers to the functions when it's meant as `to`.

Comment (duguorong009) [reply proposals/src/proposal/cosmwasm/anchor_update.rs]: Done https://github.com/webb-tools/webb-rs/pull/64/commits/1c456ad3f6d57b4b4a9a2e92c406ef7932a1f7c8

**Pull request merged**